Fix docker [WIP]

Daksh1603 opened the pull request:
Closes #28 

The following pull request deals with allowing testing of Dandisets through Docker .

Daksh1603 committed:
Changed dockerfile and added workflow

Daksh1603 committed:
Made changes to shell file

Daksh1603 committed:
Made changes to shell file

Daksh1603 committed:
Made changes to shell file

Daksh1603 committed:
Made changes

Daksh1603 committed:
Made changes

Daksh1603 committed:
Made changes

Daksh1603 committed:
Changed the shell file to build dind

Daksh1603 committed:
Mounting the docker sock

Daksh1603 committed:
Added different entry points for dind container

Daksh1603 committed:
Made changes to shell file

Daksh1603 committed:
Made changes to the shell file

Daksh1603 committed:
Removed running nwbe container and only trying updating the readme for test

Daksh1603 committed:
Added new workflow stage for test and changed dir mount

sanjayankur31 commented:
Sounds good. What error are you seeing? Could you paste the error message and maybe Padraig or I have some idea on what's going on?

Daksh1603 committed:
Implementing final nwb_table_readme changes and adding new stage for workflow for test.

Daksh1603 commented:
ERROR: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?

A workaround for this was mounting the host docker.sock to that of the containers.

As for the recent commits, the tests up till running 'generate readme' and running the nwbe container inside the DANDIArchiveShowcase container are working fine.
https://github.com/Daksh1603/DANDIArchiveShowcase/actions/runs/5500841561

It's only the final changes that are pending which are related to testing it on a limited no. of dandisets and possibly tests for other scripts too.

Daksh1603 committed:
Corrected the error in workflow

Daksh1603 committed:
Very minimal mistake rectified

Daksh1603 committed:
Changes to workflow

Daksh1603 commented:
Some additional pointers ->
1. Adding lines to generateDocker.sh : This is regarding why I added docker build commands to this shell file in particular . I tried incorporating this inside the dockerfile itself , but for a DIND container to work and essentially run, involves passing the --privileged flag to the container . This allows the container to incorporate all the capabilities of the host in order to build containers inside . However, when building a dockerfile there is no means of incorporating such privileges . After looking on ways if there is any remote support of incorporating this , I found this for [reference](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/reference.md#run---securityinsecure) . This might certainly allow the following with some additional lines in the dockerfile but the disadvantages of the following are : 
+ The user will have to setup and use a different buildkit for docker that allows this to work .
+ Not recommend for CI/CD pipelines .
As an alternative : I Just added these lines to the generateDocker.sh which works well !

2. Mounting the hosts docker.sock : This is a problem that I didn't encounter on my local machine but on github actions . I already mentioned this issue above and the workaround .

3. Persistent data generated for validation is reflected on the host machine as well which is great as any tests run inside the dandiarchive container ( along with the compability_test running inside the nwbe container which itself is inside the dandiarchive container ) generates its result on the host machine DANDIArchiveShowcase folder .

Changes yet to be made : 

Tests are running well , containers are working fine when set up but somehow the compatability_test.py file is not being discovered when testing. This seems like the main point but it would just be a little error that can be fixed within no time . I am trying to troubleshoot this but I am running into this error on my host machine : 

failed to solve with frontend dockerfile.v0: failed to create LLB definition: failed to authorize: rpc error: code = Unknown desc = failed to fetch anonymous token: Get https://auth.docker.io/token?scope=repository%3Alibrary%2Fjava%3Apull&service=registry.docker.io: net/http: TLS handshake timeout

This does happen sometimes when I am working on this, for some reason for like a day in between, when I try to run basic apk update commands, its stuck on cannot fetch try again later . This is able to work after some time, but I am convinced its my ISP acting up sometimes . As soon as I am able to build the container again on my local machine, I would resolve this in no time . As for the other tests, they seem to pass.

Daksh1603 committed:
Trying tests without Docker

Daksh1603 committed:
Minimal errors rectified

Daksh1603 committed:
yaml file errors

Daksh1603 committed:
NWBExplorer install error rectified

Daksh1603 committed:
test_docker flag fixed and working

Daksh1603 committed:
Workflow errors rectified

Daksh1603 committed:
Changes to workflow file

Daksh1603 committed:
Changes to workflow

Daksh1603 committed:
Linking /tmp

Daksh1603 committed:
Made new testing suite and added debugging statements

Daksh1603 committed:
Fixed save_folder directory,added old test and generated README uptil dandiset 350.

Daksh1603 commented on .github/workflows/ci-conda.yml:
--npm-skip will be removed .

Daksh1603 committed:
Regenerated README completely

Daksh1603 committed:
Limited Dandisets test range modified

Daksh1603 committed:
Adding time limit to resolve suspected Error dandisets

pgleeson commented on nwb_table_readme.py:
```diff
@@ -20,6 +20,21 @@
+
+class TimeoutException(Exception): pass
+
+@contextmanager
```
@Daksh1603 Has this been successfully tested on your side? Checking again here and it still gets stuck around Dandiset 13 with a memory leak, lasting longer than 5 mins.

<img width="1641" alt="Screenshot 2023-08-08 at 12 54 47" src="https://github.com/OpenSourceBrain/DANDIArchiveShowcase/assets/1556687/ab9931e6-eb41-45ba-a16a-ec736b01b705">

pgleeson commented:
@Daksh1603 Any update on the docker functionality since we spoke last week, e.g. blacklisting of problematic dandisets?

Daksh1603 committed:
Possible fix for dandisets

Daksh1603 committed:
Splitting dandiset installation

Daksh1603 committed:
Adding memory limits

Daksh1603 committed:
Fixed buffer issues and compatibility_test process.

Daksh1603 commented:
Hi @pgleeson @sanjayankur31,
Made some changes . The memory issue is no longer being reflected on my machine and hopefully the github actions test can also reflect the same. Please try it out and let me know if the problem still persists .

Daksh1603 committed:
Minor Syntax Fix
